### PR TITLE
Use Kokkos built-in reducer in bounding scene calculation

### PR DESCRIPTION
**Pull request by dalg24**
Prefer built-in reducer with user-defined scalar type to custom reducer because it is not currently supported with the Kokkos OpenMPTarget backend.

I was able to pass all examples, benchmarks, and tests with these changes (modulo minor change to tools annotation check because one allocation was not prefixed properly in Kokkos).  I did not enable MPI.

This is only a draft because I wanted to clarify requirements on the scalar type to work with build-in reducers.

### CODE REVIEW

**Review by aprokop — COMMENTED** [review]
Need to run performance benchmark for construction

File: src/details/ArborX_Box.hpp [hunk 1]
```diff
@@ -65,6 +65,55 @@ struct Box
...
+    for (int d = 0; d < 3; ++d)
+    {
+      minCorner()[d] = min(minCorner()[d], point[d]);
+      maxCorner()[d] = max(maxCorner()[d], point[d]);
+    }
+    return *this;
+  }
+
+private:
+  friend KOKKOS_FUNCTION Box operator+(Box box, Point const &point)
```
[the remark this answers is not in the document]
Comment (dalg24) [reply src/details/ArborX_Box.hpp]: You commented on something I removed but the same pattern is to be found below for the binary plus operator with two boxes as operands.
The alternative is
```C++
friend KOKKOS_FUNCTION Box operator+(Box const &box, Box const &other) {
  Box tmp{box};
  return tmp += other;
}
```
I wrote less and I expect the compiler generates the same code for it.

I don't think it is any less performant but we need to measure.

File: src/details/ArborX_DetailsTreeConstruction.hpp [hunk 2]
```diff
@@ -23,43 +23,22 @@
...
 
+namespace Kokkos
+{ // reduction identity must be defined in Kokkos namespace
+template <>
+struct reduction_identity<ArborX::Box>
+{
+  KOKKOS_FUNCTION static ArborX::Box sum() { return {}; }
+};
+} // namespace Kokkos
+
```
[the remark this answers is not in the document]
Comment (dalg24) [reply src/details/ArborX_DetailsTreeConstruction.hpp:34]: I moved it closer to its point of use.  I don't want to have to include anything else than `<Kokkos_Macros.hpp>` in the `Box` header and a forward declaration in the `Kokkos::` namespace sounds silly to me.

File: src/details/ArborX_Box.hpp [hunk 3]
```diff
@@ -65,6 +65,55 @@ struct Box
...
+  {
+    return box += point;
+  }
+
+  friend KOKKOS_FUNCTION Box operator+(Box box, Box const &other)
```
[the remark this answers is not in the document]
Comment (dalg24) [reply src/details/ArborX_Box.hpp:110]: I think it is bringing too much attention to a detail.

File: src/details/ArborX_Box.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (aprokop) [reply src/details/ArborX_Box.hpp]: Sorry, the comment was from a draft review. In `Box operator+(Box, Box const&)`, are there going to be 2 `Box` copies? Or is the second one going to be through return type optimization?

File: src/details/ArborX_DetailsTreeConstruction.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (aprokop) [reply src/details/ArborX_DetailsTreeConstruction.hpp:34]: Essentially, you are expecting that the `Box` reduction is only going to happen in the tree construction. We'll probably move out scene bounding box calculation out of this file in the future anyway, as it is needed in kd-tree too, as well as anything using Morton codes. Find to leave for now.

File: src/details/ArborX_Box.hpp [hunk 4]
```diff
@@ -65,6 +65,55 @@ struct Box
...
+  {
+    using KokkosExt::max;
+    using KokkosExt::min;
+
+    for (int d = 0; d < 3; ++d)
+    {
+      minCorner()[d] = min(minCorner()[d], other.minCorner()[d]);
+      maxCorner()[d] = max(maxCorner()[d], other.maxCorner()[d]);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (dalg24) [reply src/details/ArborX_Box.hpp:92]: You can look at how we were doing with `expand(Box, Box)`

[the remark this answers is not in the document]
Comment (aprokop) [reply src/details/ArborX_Box.hpp:92]: But that one allowed both volatile and non-volatile argument. Here, the function itself is volatile. So I'm not sure what you mean.

**Comment by aprokop** [comment]
I ran comparison master (9338ba9) with branch (c031119) on Summit.
I don't 100% trust these results, as in #361 there was some weirdness about CPU results, but this should be investigated further, as the current data indicates significant regression.

```
BM_construction<ArborX::BVH<Serial>>/10000/0/manual_time_median                  +0.1385         +0.1365          1811          2062          1880          2137
BM_construction<ArborX::BVH<Serial>>/100000/0/manual_time_median                 +0.0178         +0.0178         16180         16467         16254         16543
BM_construction<ArborX::BVH<Serial>>/1000000/0/manual_time_median                +0.0013         +0.0013        176128        176349        176616        176846
BM_construction<ArborX::BVH<Serial>>/10000/1/manual_time_median                  +0.1209         +0.1200          1786          2002          1855          2077
BM_construction<ArborX::BVH<Serial>>/100000/1/manual_time_median                 +0.0123         +0.0126         15902         16097         15970         16172
BM_construction<ArborX::BVH<Serial>>/1000000/1/manual_time_median                -0.0026         -0.0025        173041        172596        173523        173088
BM_construction<ArborX::BVH<OpenMP>>/10000/0/manual_time_median                  +1.2985         +1.3573           317           729           327           771
BM_construction<ArborX::BVH<OpenMP>>/100000/0/manual_time_median                 +0.2386         +0.2528          1537          1904          1552          1944
BM_construction<ArborX::BVH<OpenMP>>/1000000/0/manual_time_median                +0.0171         +0.0179         14855         15110         15390         15665
BM_construction<ArborX::BVH<OpenMP>>/10000/1/manual_time_median                  +1.2187         +1.2767           331           734           341           776
BM_construction<ArborX::BVH<OpenMP>>/100000/1/manual_time_median                 +0.1950         +0.2082          1725          2061          1739          2102
BM_construction<ArborX::BVH<OpenMP>>/1000000/1/manual_time_median                +0.0141         +0.0148         17538         17786         18073         18340
BM_construction<ArborX::BVH<Cuda>>/10000/0/manual_time_median                    +0.0059         +0.0007          1411          1420          1424          1425
BM_construction<ArborX::BVH<Cuda>>/100000/0/manual_time_median                   -0.0012         -0.0005          2159          2156          2551          2550
BM_construction<ArborX::BVH<Cuda>>/1000000/0/manual_time_median                  -0.0045         +0.0002          6961          6930          7406          7407
BM_construction<ArborX::BVH<Cuda>>/10000/1/manual_time_median                    -0.0018         +0.0008          1422          1419          1424          1425
BM_construction<ArborX::BVH<Cuda>>/100000/1/manual_time_median                   -0.0009         -0.0003          2159          2157          2552          2551
BM_construction<ArborX::BVH<Cuda>>/1000000/1/manual_time_median                  -0.0001         -0.0001          6919          6919          7408          7408
```

**Comment by aprokop** [comment]
I've rerun the branch commit 3 additional times. All of them were on par with master. I'm not sure what changed on Summit in the past couple months, it was really stable before that. I'll try running different commits in different job submissions in the future.

**Review by aprokop — APPROVED** [review]
LGTM

**Comment by aprokop** [comment]
The build failed for 10.2-NVCC, but it just ran out of space. I'm confident this is not a problem, so I'm merging as it is.

**Pull request merged**